renku_notebooks/api/schemas.py

```diff
@@ -248,66 +256,133 @@ class Meta:
+                # INFO: the command is found but cannot be invoked
+                126: "The command to start the server cannot be invoked. "
+                "Please make sure your Dockerfile is correct and up-to-date.",
+                # INFO: the command cannot be found at all
+                127: "The image does not contain the required command to start the server. "
+                "Please make sure your Dockerfile is correct and up-to-date.",
+                # INFO: the container exited with an invalid exit code
+                # happens when container fully runs out of storage
+                128: "The server shut down unexpectedly. Please ensure"
+                "that your Dockerfile is correct and up-to-date. "
+                "In some cases this can be the result of low disk space, "
+                "please restart your server with more storage.",
+                # INFO: the container aborted itself using the abort() function.
+                134: default_server_error_message,
+                # INFO: receiving SIGKILL - eviction or oomkilled should trigger this
+                137: "The server was terminated by the cluster. Potentially because of "
+                "consuming too much resources. Please restart your server and request "
+                "more memory and storage.",
+                # INFO: segmentation fault
+                139: default_server_error_message,
+                # INFO: receiving SIGTERM
+                143: default_server_error_message,
+            }
+            return exit_code_msg_xref.get(exit_code, default_server_error_message)
+
+        def get_failed_message(failed_containers):
+            """The failed message tries to extract a meaningful error info from the containers."""
+            num_failed_containers = len(failed_containers)
+            if num_failed_containers == 0:
+                return None
+            for container in failed_containers:
+                exit_code = get_failed_container_exit_code(container)
+                container_name = container.get("name", "Unknown")
+                if (
+                    container_name == "git-clone" and exit_code == 128
+                ) or container_name == "jupyter-server":
+                    # INFO: The git-clone init container ran out of disk space
+                    # or the server container failed
+                    user_correctable_message = get_user_correctable_message(exit_code)
+                    return user_correctable_message
+            return (
+                f"There are failures in {num_failed_containers} auxiliary "
+                "server containers. Please restart your session as this may be "
+                "an intermittent problem. If issues persist contact your "
+                "administrator or the Renku team."
+            )
+
+        def get_all_container_statuses(js):
+            return js["status"].get("mainPod", {}).get("status", {}).get(
```

Comment:
Not really something for this PR, but did we ever consider using something like JSONPath/XPath for getting nested dict values? I see a lot of .get(..., {}).get(..., {}) style calls in renku_notebooks and I think that could make a lot of things cleaner.

Like https://pypi.org/project/jmespath/ , https://pypi.org/project/dpath/ or https://github.com/kennknowles/python-jsonpath-rw (looks quite nice and is semi-active but hasn't [had an official release for a while](https://github.com/kennknowles/python-jsonpath-rw/issues/98) :disappointed:  )

Reply:
That is a great idea and I think we should use something like that. Because whenever I did not do `.get(..., {}).get(..., {})` I regretted it afterwards and had to fix bugs. I will open an issue.

Reply:
https://github.com/SwissDataScienceCenter/renku-notebooks/issues/945